# Networks: structure, evolution & processes
**Internet Analytics - Lab 2**

---

**Group:** K

**Names:**

*  Xavier Jeanmonod
*  Adrian Baudat
*  Simon Wicky

---

#### Instructions

*This is a template for part 1 of the lab. Clearly write your answers, comments and interpretations in Markodown cells. Don't forget that you can add $\LaTeX$ equations in these cells. Feel free to add or remove any cell.*

*Please properly comment your code. Code readability will be considered for grading. To avoid long cells of codes in the notebook, you can also embed long python functions and classes in a separate module. Don’t forget to hand in your module if that is the case. In multiple exercises, you are required to come up with your own method to solve various problems. Be creative and clearly motivate and explain your methods. Creativity and clarity will be considered for grading.*

---

## 2.1 Exploration of real networks

#### Exercise 2.1

In [1]:
import networkx as nx
import numpy as npy

In [2]:
Data = open('../data/network1.csv', "r")
next(Data,None)
Graphtype = nx.Graph()
G = nx.parse_edgelist(Data, delimiter=',', create_using=Graphtype,
                      nodetype=int)
print("Number of edges : " + str(len(G.edges)))
print("Number of nodes : " + str(len(G.nodes)))

Number of edges : 18584
Number of nodes : 13033


#### Exercise 2.2:  Node degree distribution

In [1]:
import matplotlib.pyplot as plt
import math
nodesDeg = G.degree()
degrees = [d for _, d in nodesDeg]
#print(degrees)
count = [()]
for d in degrees:
    if d not in count:
        count.append((d, 1))
    else: 
        count[d] += 1
count.pop(0)
numberOf = [v for _, v in count]
#print(numberOf)
    
plt.plot(degrees, numberOf)
plt.xscale('log')
plt.yscale('log')
plt.show()

#Should correct this

NameError: name 'G' is not defined

#### Exercise 2.3: Giant component

In [4]:
print("Number of connected components : " + str(nx.number_connected_components(G)))
size = [len(o) for o in nx.connected_components(G)]

print("Histogram of connected components (size : number of CC)")
for i in range(max(size) + 1):
    if size.count(i) > 0:
        print(str(i) + " : " + str(size.count(i)))
    

Number of connected components : 192
Histogram of connected components (size : number of CC)
2 : 149
3 : 29
4 : 7
5 : 3
6 : 1
7 : 1
16 : 1
12576 : 1


Looking at the data above, we can see that there is 192 connected components, with one giant components of size 12756

#### Exercise 2.4: Short paths & Small-worlds

In [5]:
def numberOfNodesHops(start, r, G):
    neighbors = set([start])
    for i in range(r):
        neighbors = set([neighbor for node in neighbors \
                     for neighbor in G[node]])
    return neighbors

from random import choice
giant_cc = [e for e in nx.connected_component_subgraphs(G) if len(e.nodes()) == max(size)][0]
random_node = choice(list(giant_cc.nodes))
len(numberOfNodesHops(random_node, 7, giant_cc))

12575

The average shortest path is computed as follow :
*    Pick one node at random, and computes its average shortest path length with every other node
*    Do step 1 on 1000 nodes and average the whole

This computation seems relatively representative of the average shortest path length, the real computation is too costly.

And so this computation yields 4.174

#### Exercise 2.5: Network comparison

In [6]:
Data2 = open('../data/network2.csv', "r")
next(Data2,None)
Graphtype = nx.Graph()
G2 = nx.parse_edgelist(Data2, delimiter=',', create_using=Graphtype,
                      nodetype=int)
print("Number of edges : " + str(len(G2.edges)))
print("Number of nodes : " + str(len(G2.nodes)))

Number of edges : 43516
Number of nodes : 26542


In [7]:
print("Number of connected components : " + str(nx.number_connected_components(G2)))
size2 = [len(o) for o in nx.connected_components(G2)]

print("Histogram of connected components (size : number of CC)")
for i in range(max(size2) + 1):
    if size2.count(i) > 0:
        print(str(i) + " : " + str(size2.count(i)))

Number of connected components : 5
Histogram of connected components (size : number of CC)
2 : 1
5 : 1
24 : 1
30 : 1
26481 : 1


In [8]:
def numberOfNodesHops(start, r, G):
    neighbors = set([start])
    for i in range(r):
        neighbors = set([neighbor for node in neighbors \
                     for neighbor in G[node]])
    return neighbors

from random import choice
giant_cc = [e for e in nx.connected_component_subgraphs(G2) if len(e.nodes()) == max(size2)][0]
random_node = choice(list(giant_cc.nodes))
len(numberOfNodesHops(random_node, 10, giant_cc))

54

#### Exercise 2.6: Network identification